openapi-codegen/core/src/main/scala/sttp/tapir/codegen/ClassDefinitionGenerator.scala

```diff
@@ -7,17 +7,45 @@ import sttp.tapir.codegen.openapi.models.OpenapiSchemaType._
 
 import scala.annotation.tailrec
 
+case class GeneratedClassDefinitions(classRepr: String, serdeRepr: Option[String])
+
 class ClassDefinitionGenerator {
-  val jsoniterDefaultConfig = "com.github.plokhotnyuk.jsoniter_scala.macros.CodecMakerConfig.withAllowRecursiveTypes(true).withDiscriminatorFieldName(scala.None)"
 
   def classDefs(
       doc: OpenapiDocument,
       targetScala3: Boolean = false,
       queryParamRefs: Set[String] = Set.empty,
       jsonSerdeLib: JsonSerdeLib.JsonSerdeLib = JsonSerdeLib.Circe,
-      jsonParamRefs: Set[String] = Set.empty
-  ): Option[String] = {
+      jsonParamRefs: Set[String] = Set.empty,
+      fullModelPath: String = ""
+  ): Option[GeneratedClassDefinitions] = {
     val allSchemas: Map[String, OpenapiSchemaType] = doc.components.toSeq.flatMap(_.schemas).toMap
+    val allOneOfSchemas = allSchemas.collect { case (name, oneOf: OpenapiSchemaOneOf) => name -> oneOf }.toSeq
+    val adtInheritanceMap: Map[String, Seq[String]] = allOneOfSchemas
+      .flatMap { case (name, schema) =>
+        val children = schema.types.map {
+          case ref: OpenapiSchemaRef if ref.isSchema => Right(ref.stripped)
+          case other                                 => Left(other.getClass.getName)
+        }
+        children.collect { case Left(unsupportedChild) =>
+          throw new NotImplementedError(
+            s"oneOf declarations are only supported when all variants are declared schemas. Found type '$unsupportedChild' as variant of $name"
+          )
+        }
+        val validatedChildren = children.collect { case Right(kv) => kv }
+        schema.discriminator match {
+          case None =>
+          case Some(d) =>
+            val targetClassNames = d.mapping.values.map(_.split('/').last).toSet
+            if (targetClassNames != validatedChildren.toSet)
+              throw new IllegalArgumentException(
+                s"Discriminator values $targetClassNames did not match schema variants $validatedChildren for oneOf defn $name"
+              )
+        }
+        validatedChildren.map(_ -> name)
+      }
+      .groupBy(_._1)
+      .mapValues(_.map(_._2))
```

Comment:
God I miss groupMap & groupMapReduce. Should get it sorted really. Not this time.